SCClassLibrary/Common/Math/SimpleNumber.sc

```diff
@@ -705,13 +711,24 @@ SimpleNumber : Number {
 		};
 		hours = (decimal.div(3600) % 24).asString.padLeft(2, "0").add($:);
 		minutes = (decimal.div(60) % 60).asString.padLeft(2, "0").add($:);
-		seconds = (decimal % 60).asString.padLeft(2, "0").add($.);
-
-		// min value of precision is 0.001; this ensures that we stick to 3 decimal places in the
-		// formatted string.
-		precision = max(precision, 0.001);
-		mseconds = this.frac.round(precision) * 1000;
-		mseconds = mseconds.round.asInteger.asString.padLeft(3, "0");
+		seconds = (decimal % 60).asString.padLeft(2, "0");
+		mseconds = if(decimalPlaces > 0) {
+			var curPrec = precision;
+			var curFrac = number.frac;
+			var curDecimalPlaces = decimalPlaces;
+			var str = ".";
+			// this could be simplified once sclang gains sprintf functionality, see issue #3570
+			while { curDecimalPlaces > 0 } {
+				var curInt;
+				curPrec = curPrec * 10;
+				curFrac = (curFrac * 10).round(curPrec);
+				curInt = curFrac.asInteger;
+				str = str ++ curInt.asString;
+				curFrac = curFrac - curInt;
+				curDecimalPlaces = curDecimalPlaces - 1;
```

Comment:
why do you round to `curPrec` every pass? that seems unnecessary on the face of it

Reply:
This is to compensate for lost precision/rounding error when "cutting" the `.frac` at every pass. Without this at least one test failed.